packages/web-components/fast-ssr/src/element-renderer/fast-element-renderer.ts

```diff
@@ -117,32 +118,55 @@ export abstract class SyncFASTElementRenderer extends FASTElementRenderer
 }
 export abstract class AsyncFASTElementRenderer extends FASTElementRenderer
     implements AsyncElementRenderer {
-    renderAttributes = renderAttributesAsync;
+    constructor(tagName: string, renderInfo: RenderInfo) {
+        super(tagName, renderInfo);
+
+        this.element.addEventListener("pending-task", this.pendingTaskHandler);
+    }
+    public *renderAttributes(
+        this: AsyncFASTElementRenderer
+    ): IterableIterator<string | Promise<string>> {
+        if (this.element !== undefined) {
+            if (this.awaiting.size) {
+                yield this.pauseRendering().then(() => "");
+            }
+            const { attributes } = this.element;
+
+            for (
+                let i = 0, name, value;
+                i < attributes.length && ({ name, value } = attributes[i]);
+                i++
+            ) {
+                if (value === "" || value === undefined || value === null) {
+                    yield ` ${name}`;
+                } else {
+                    yield ` ${name}="${escapeHtml(value)}"`;
+                }
+            }
+        }
+    }
     renderShadow = renderShadow as (
         renderInfo: RenderInfo
     ) => IterableIterator<string | Promise<string>>;
-}
 
-function* renderAttributesSync(this: FASTElementRenderer): IterableIterator<string> {
-    if (this.element !== undefined) {
-        const { attributes } = this.element;
-        for (
-            let i = 0, name, value;
-            i < attributes.length && ({ name, value } = attributes[i]);
-            i++
-        ) {
-            if (value === "" || value === undefined || value === null) {
-                yield ` ${name}`;
-            } else {
-                yield ` ${name}="${escapeHtml(value)}"`;
-            }
+    private async pauseRendering() {
+        for await (const awaiting of this.awaiting) {
+            continue;
         }
+
+        this.awaiting.clear();
     }
+
+    private awaiting: Set<Promise<void>> = new Set();
+
+    private pendingTaskHandler = (e: Event) => {
+        if (e instanceof PendingTaskEvent) {
```

Comment:
Should we instead check the type?

Reply:
Updated this to put a static test method on PendingTaskEvent, which gives us a good TypeScript experience while not checking instance type specifically.